nixos/lib/build-vms.nix

```diff
@@ -58,24 +58,42 @@ rec {
 
       machinesNumbered = zipLists machines (range 1 254);
 
+      qemu-common = import ./qemu-common.nix { inherit lib pkgs; };
+
       nodes_ = forEach machinesNumbered (m: nameValuePair m.fst
         [ ( { config, nodes, ... }:
             let
-              interfacesNumbered = zipLists config.virtualisation.vlans (range 1 255);
-              interfaces = forEach interfacesNumbered ({ fst, snd }:
-                nameValuePair "eth${toString snd}" { ipv4.addresses =
-                  [ { address = "192.168.${toString fst}.${toString m.snd}";
+              # Convert legacy VLANs to named interfaces and merge with explicit interfaces.
+              vlansNumbered = forEach (zipLists config.virtualisation.vlans (range 1 255)) (v: {
+                name = "eth${toString v.snd}";
+                vlan = v.fst;
+                assignIP = true;
+              });
+              explicitInterfaces = mapAttrsToList (n: v: v // { name = n; }) config.virtualisation.interfaces;
+              interfaces = vlansNumbered ++ explicitInterfaces;
+              interfacesNumbered = zipLists interfaces (range 1 255);
+
+              # Automatically assign IP addresses to requested interfaces.
+              assignIPs = filter (i: i.assignIP) interfaces;
+              ipInterfaces = forEach assignIPs (i:
+                nameValuePair i.name { ipv4.addresses =
+                  [ { address = "192.168.${toString i.vlan}.${toString m.snd}";
                       prefixLength = 24;
                   } ];
                 });
 
+              qemuOptions = flatten (forEach interfacesNumbered ({ fst, snd }:
+                qemu-common.qemuNICFlags snd fst.vlan m.snd));
+              udevRules = forEach interfacesNumbered ({ fst, snd }:
+                "SUBSYSTEM==\"net\",ACTION==\"add\",ATTR{address}==\"${qemu-common.qemuNicMac fst.vlan m.snd}\",NAME=\"${fst.name}\"");
```

Comment:
I chose to use udev rules rather than `systemd.network.links` because in the latter case, systemd ignores all but one file that applies to a given interface. I did not want to preclude users from using that module elsewhere.